### PR TITLE
store/tikv: pipe the split and scatter task

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?

close https://github.com/pingcap/tidb/issues/22969

Problem Summary:
Splitting and Scattering at the same time will cause multiple conflicts on the same region.
image

### What is changed and how it works?

pipeline all splits task and scatter tasks

### Release note 

- Make the time consuming of `split table` more stable 
```
I have run multiple times  to split 4000 regions, and the time is stale about `1m30s`
2021/03/17 17:25:57 split single sql result: [map[SCATTER_FINISH_RATIO:1 TOTAL_SPLIT_REGION:2001]]
2021/03/17 17:25:57 split all sql run total cost: 1m28.908304306s
2021/03/17 17:25:57 split all sql run success!!!

```

### CODE REVIEW

File: store/tikv/split_region.go [hunk 1]
```diff
@@ -157,29 +192,41 @@ func (s *KVStore) batchSendSingleRegion(bo *Backoffer, batch batch, scatter bool
...
-	for i, r := range spResp.Regions {
+func (s *KVStore) scatterRegionsInBatchSplitResp(bo *Backoffer, tableID *int64, batchResp singleBatchResp) singleBatchResp {
+	var err error
+	start := time.Now()
+	spResp := batchResp.resp.Resp.(*kvrpcpb.SplitRegionResponse)
+	for _, r := range spResp.Regions {
+		tid := int64(0)
+		if tableID != nil {
+			tid = *tableID
+		}
```
Comment (crazycs520) [inline store/tikv/split_region.go:213]:
```suggestion
	tid := int64(0)
	if tableID != nil {
		tid = *tableID
	}
	for _, r := range spResp.Regions {
```

File: store/tikv/split_region.go [hunk 2]
```diff
@@ -188,6 +235,15 @@ func (s *KVStore) batchSendSingleRegion(bo *Backoffer, batch batch, scatter bool
 			break
 		}
 	}
+	tid := int64(0)
+	if tableID != nil {
+		tid = *tableID
+	}
```
Comment (crazycs520) [inline store/tikv/split_region.go:241]:
```suggestion
```

File: store/tikv/split_region.go [hunk 3]
```diff
@@ -157,28 +199,41 @@ func (s *KVStore) batchSendSingleRegion(bo *Backoffer, batch batch, scatter bool
...
+func (s *KVStore) scatterRegionsInBatchSplitResp(bo *Backoffer, tableID *int64, batchResp singleBatchResp) singleBatchResp {
+	var err error
+	start := time.Now()
+	tid := int64(0)
+	if tableID != nil {
+		tid = *tableID
+	}
+	spResp := batchResp.resp.Resp.(*kvrpcpb.SplitRegionResponse)
+	for _, r := range spResp.Regions {
 		if err = s.scatterRegion(bo, r.Id, tableID); err == nil {
```
Comment (jackysp) [inline store/tikv/split_region.go:222]: So it's still using https://github.com/pingcap/tidb/blob/579421f6e02ab5cdde3344aa25815725a1561f12/store/tikv/split_region.go#L217 the batch API to scatter regions one by one. Which may make the scatter not work?

Comment (jackysp) [reply store/tikv/split_region.go:222]: I mean https://github.com/pingcap/tidb/issues/22788 .
OK, I'm off topic.

Comment (Yisaer) [reply store/tikv/split_region.go:222]: The current region scatter may not work in a specific case which described in https://github.com/tikv/pd/pull/3422 ,  I think it's not related to the api.

**Pull request closed without merging**